spotbugs/src/main/java/edu/umd/cs/findbugs/FindBugs2.java

```diff
@@ -1062,33 +1095,40 @@ private void analyzeApplication() throws InterruptedException {
                     currentAnalysisContext.setClassBeingAnalyzed(classDescriptor);
 
                     try {
-                        for (Detector2 detector : detectorList) {
-                            if (Thread.interrupted()) {
-                                throw new InterruptedException();
-                            }
-                            if (isHuge && !FirstPassDetector.class.isAssignableFrom(detector.getClass())) {
-                                continue;
-                            }
-                            if (DEBUG) {
-                                System.out.println("Applying " + detector.getDetectorClassName() + " to " + classDescriptor);
-                                // System.out.println("foo: " +
-                                // NonReportingDetector.class.isAssignableFrom(detector.getClass())
-                                // + ", bar: " + detector.getClass().getName());
+                        Collection<Callable<Void>> tasks = Arrays.stream(detectorList).map(detector ->
```

Comment:
sorry, @KengoTODA don't you need to write thread-safe AnalysisCache for it too?
What will happen if 2 detectors ask for the same analysis from cache?

Reply:
Yes you're right. As documented in the description of this PR, this change isn't complete to run detectors in parallel.

Reply:
> Yes you're right. As documented in the description of this PR, this change isn't complete to run detectors in parallel.

do you  have any idea how to finish this task with situation about BCEL? as I understand moving to  ASM is not really possible in reality...so how to solve it? I think it's the main problem about it.

Reply:
Sorry to necro, I'm wondering about the reasons it is not easy to migrate from BCEL to ASM. Are there any docs regarding why this isn't possible?

Reply:
> Sorry to necro, I'm wondering about the reasons it is not easy to migrate from BCEL to ASM.

It is simple: these API's are completely different - one id DOM based, another one is visitor based (the "fast" ASM API), so porting BCEL based code to ASM is almost a full rewrite. Of course one could use ASM tree API that is somewhat similar to BCEL code except missing ConstantPool access, but as far as I remember we will still need a replacement for class repository from BCEL that of course doesn't exists at all in ASM API.